proposals/params-span.md

```diff
@@ -0,0 +1,138 @@
+# `params Span<T>` and implicit allocation of arrays
+
+## Summary
+Avoid heap allocation for implicit allocation of arrays in specific scenarios such as `params` arguments.
+
+## Motivation
+`params` array parameters provide a convenient way to call a method that takes an arbitrary length list of arguments.
+However, using an array type for the parameter means the compiler must implicitly allocate an array on the heap at each call site.
+
+If we extend `params` types to include the `ref struct` types `Span<T>` and `ReadOnlySpan<T>`, where values of those types cannot escape the call stack, the array at the call site may be created on the stack instead.
+
+And if we're extending `params` to other types, we  could also allow `params IEnumerable<T>` to avoid allocating and copying collections at call sites that have an `IEnumerable<T>` rather than `T[]`.
+
+The benefits of `params ReadOnlySpan<T>` and `params Span<T>` are primarily for new APIs. Existing commonly used APIs such as `Console.WriteLine()` and `StringBuilder.AppendFormat()` already have overloads that avoid array allocations for common cases.
+```csharp
+public static class Console
+{
+    public static void WriteLine(string value);
+    public static void WriteLine(string format, object arg0);
+    public static void WriteLine(string format, object arg0, object arg1);
+    public static void WriteLine(string format, object arg0, object arg1, object arg2);
+    public static void WriteLine(string format, params object[] arg);
+}
+```
+
+## Detailed design
+
+### Extending `params`
+`params` parameters will be supported with types `Span<T>`, `ReadOnlySpan<T>`, and `IEnumerable<T>`.
+
+A call in _expanded_ form to a method with a `params T[]` or `params IEnumerable<T>` parameter will result in an array `T[]` allocated on the heap.
+
+A call in _expanded_ form to a method with a `params ReadOnlySpan<T>` or `params Span<T>` parameter will result in an array `T[]` created on the stack if the `params` array has no more than 8 arguments (an arbitrary limit).
```

Comment:
do we want the actual language spec to state the limit?  or do we want to leave it to be a compiler/runtime/user possible heuristic?

Reply:
I am wary of arbitrary boundaries that users don't know when they've crossed. Perhaps this is fine for the "simple expanded call" scenario, where getting the stack allocation is just a bonus, but other users will want to be able to make a call that is "stack allocate this or error", i.e. it won't accidentally cross the threshold as the code evolves. I think this is an important reason to have an "explicit" equivalent to the expanded call.

That said, it occurs to me that this was probably discussed in an LDM that I missed, so this might have all been touched on already..

Reply:
Created an open issue for heuristics. To explicitly create the span on the stack, we should consider allowing `stackalloc [] { x, y, z }`.